### PR TITLE
Cognito, S3, DyanmoDB Lambda Trigger support

### PR DESCRIPTION
**Pull request by kaustavghosh06**
*Description of changes:*
Cognito, S3, DyanmoDB Lambda Trigger support.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-category-auth/index.js [hunk 1]
```diff
@@ -188,12 +188,25 @@ async function checkRequirements(requirements, context) {
 async function initEnv(context) {
   const { amplify } = context;
   const { resourcesToBeCreated, resourcesToBeDeleted, resourcesToBeUpdated } = await amplify.getResourceStatus('auth');
+  let toBeCreated = [];
+  let toBeDeleted = [];
+  let toBeUpdated = [];
 
-  resourcesToBeDeleted.forEach((authResource) => {
+  if (resourcesToBeCreated && resourcesToBeCreated.length > 0) {
```
Comment (kaustavghosh06) [inline packages/amplify-category-auth/index.js]: Why do we need to do this additonal filtering? The amplify.getResourceStatus('auth') function should already filter your resources based on auth.

File: packages/amplify-category-auth/provider-utils/awscloudformation/assets/cognito-defaults.js [hunk 2]
```diff
@@ -29,20 +29,22 @@ const roles = {
...
   return ({
+    resourceNameTruncated: `${projectName.substring(0, 6)}${sharedId}`,
     userPoolName: `${projectName}_userpool_${sharedId}`,
     autoVerifiedAttributes: ['email'],
     mfaConfiguration: 'OFF',
     mfaTypes: ['SMS Text Message'],
-    roleName: `${projectNameTruncated}_sns-role`,
-    roleExternalId: `${projectNameTruncated}_role_external_id`,
-    policyName: `${projectNameTruncated}-sns-policy`,
+    // roleName: `${projectNameTruncated}_sns-role`,
```
Comment (kaustavghosh06) [inline packages/amplify-category-auth/provider-utils/awscloudformation/assets/cognito-defaults.js]: Why are these commented out?

File: packages/amplify-graphql-docs-generator/package.json [hunk 3]
```diff
@@ -38,13 +38,16 @@
     "typescript": "^3.5.1"
   },
   "dependencies": {
+    "aws-sdk": "^2.475.0",
```
Comment (kaustavghosh06) [inline packages/amplify-graphql-docs-generator/package.json]: Are these graphql transformer package changes your changes?

File: packages/amplify-category-auth/commands/auth/push.js [hunk 4]
```diff
@@ -8,6 +9,9 @@ module.exports = {
     const resourceName = parameters.first;
     context.amplify.constructExeInfo(context);
     return amplify.pushResources(context, category, resourceName)
+      .then(async () => {
+        await uploadFiles(context);
```
[the remark this answers is not in the document]
Comment (haverchuck) [reply packages/amplify-category-auth/commands/auth/push.js]: This is for uploading files to the s3 static site.

File: package.json [hunk 5]
```diff
@@ -41,6 +41,7 @@
   "author": "Amazon Web Services",
   "license": "Apache-2.0",
   "dependencies": {
+    "cloudform": "^4.0.0-alpha.0",
```
[the remark this answers is not in the document]
Comment (haverchuck) [reply package.json]: Hmmm... I think I introduced this when I was trying to reserve a build error.  It is probably solved by your rebasing.  I'll remove and test.

File: packages/amplify-category-analytics/package.json [hunk 6]
```diff
@@ -9,6 +9,7 @@
     "fs-extra": "^7.0.0",
     "inquirer": "6.3.1",
     "opn": "^5.3.0",
+    "ora": "^3.4.0",
```
[the remark this answers is not in the document]
Comment (haverchuck) [reply packages/amplify-category-analytics/package.json]: Yes - meant to only add this to auth, but it got added here for some reason.  I'm removing it.

File: packages/amplify-category-auth/index.js (hunk 1 above)
Comment (haverchuck) [reply packages/amplify-category-auth/index.js]: That's what I would exect as well, but for some reason I'm getting this:
```
resourcesToBeCreated
Array(2) [Object, Object]
length:2
__proto__:Array(0) [, …]
0:Object {service: "Cognito", providerPlugin: "awscloudformation", dependsOn: Array(1), …}
category:"auth"
dependsOn:Array(1) [Object]
providerPlugin:"awscloudformation"
resourceName:"clilaminit8f1d23e0"
service:"Cognito"
__proto__:Object {constructor: , __defineGetter__: , __defineSetter__: , …}
1:Object {service: "Lambda", providerPlugin: "awscloudformation", build: true, …}
build:true
category:"function"
providerPlugin:"awscloudformation"
resourceName:"clilaminit8f1d23e0PostConfirmation"
service:"Lambda"
__proto__:Object {constructor: , __defineGetter__: , __defineSetter__: , …}
```

I wasn't entirely sure what the impacts of changing getResourceStatus would be, so I thought it safer to do filtering on the auth side.

Comment (kaustavghosh06) [reply packages/amplify-category-auth/index.js]: It already does the filtering for you..checkout the function and the params it expects.

File: packages/amplify-category-auth/provider-utils/awscloudformation/assets/cognito-defaults.js (hunk 2 above)
Comment (haverchuck) [reply packages/amplify-category-auth/provider-utils/awscloudformation/assets/cognito-defaults.js]: I commented them out while testing their removal (had to decrease the number of parameters in the cf template, and these never change anyway so they aren't needed).  I'm deleting them now.  Thanks for catching this.

File: packages/amplify-graphql-docs-generator/package.json (hunk 3 above)
Comment (haverchuck) [reply packages/amplify-graphql-docs-generator/package.json]: no - they should be reverted now.

**Pull request merged**